### DISCUSSION TITLE
Missing Pieces

### DISCUSSION THREAD

**Discussion by phronmophobic** [phronmophobic/membrane#66] [General]
The focus of this discussion to have a place to track "missing pieces". A missing piece is any improvement that could made to improve the workflow for creating user interfaces with membrane. There's no particular format and it's ok if what's "missing" is a little fuzzy. The goal is really just to collect bits of feedback, and hopefully, find a way to turn that into improvements to code, documentation, tutorials, etc.

**Comment by phronmophobic** [comment]
One idea based on following @realgenekim's work is to provide some sort of dev-mode for membrane that offers a suite of tools during development. Some general examples:

- bounding boxes for elements
- automatically catch exceptions in view function and display
- options for automatically refreshing various caches.

**Comment by phronmophobic** [comment]
Based on @jlmr's comments (https://clojurians.slack.com/archives/CVB8K7V50/p1677005660717379?thread_ts=1676886695.956009&cid=CVB8K7V50). It might be nice to have some clj-kondo configs or similar that show warnings for common mistakes.

**Comment by phronmophobic** [comment]
More utilities to support grid based layouts (see https://clojurians.slack.com/archives/CVB8K7V50/p1676317261261979).

**Reply by jjttjj** [reply]
To summarize briefly, I was looking to lay out elements in a grid and then draw lines connecting those elements, resulting in a graph, something like graphviz would produce. I happened to want my elements laid out in a grid, so the result would be for example, 4 rows by 4 columns or rectangles, with each rectangle having a line connecting to a random one of its neighbors.

Ultimately I'm not sure that handling this type of usage under the scope of a general grid layout is best, since I'd imagine a grid layout would mostly be meant for UIs where elements probably wouldn't need or want some form of access to the low level coordinate info of all other cells in the grid.

On slack you proposed a solution that made a grid element with the normal `ui/horizontal-laout`/`vertical-layout` functions and then walked that grid to determine the coordinates of its children (as I initially asked, thanks for that!).

Eventually I went with a version that instead just had a function mapping my domain objects (the things represented in the graph) to an x/y coordinate based on the overall set of objects, and then using this to lay out the elements themselves and the lines connecting them. I like to use this library to help with the scale transformations https://github.com/scicloj/wadogo

I wonder if there is some abstract element type to be captured here, I'm not really convinced there is. But I will keep experimenting with it (probably slowly). I have vague thoughts about a special child type that will "bubble up" to be collected into a key of a special parent type.

**Reply by phronmophobic** [reply]
Interesting. Thanks for the summary 👍

**Comment by phronmophobic** [comment]
Scrolling sensitivity and direction ignores user settings. It might also be a good idea to offer better tools for devs to tweak scrolling sensitivity.

See discussion:

https://clojurians.slack.com/archives/CVB8K7V50/p1678438796441739
